.eslintrc.js

```diff
@@ -8,6 +8,8 @@ module.exports = {
   globals: {
     Atomics: 'readonly',
     SharedArrayBuffer: 'readonly',
+    i18next: true,
```

Comment:
I'm not quite aware of I18next, as I primarily work in Ruby. Could you maybe elaborate a few ways in which this is superior to i18n-js, I'm genuinely curious and would like to learn more about this, thanks!

Reply:
Thank you for reviewing. Maybe I should rather start from how i18n-js is not ideal. It is ultimately about how the vanilla Rails I18n (from which it is ported) is featureless. Much to my surprise, it can only handle languages equally or less inflective as English out of the box. It only does decent job worldwide because of a number of extensions such as [rails-i18n](/svenfuchs/rails-i18n). That means, without them, you have to redefine pluralization rules and other things yourself for any language not conforming this pattern, [as is written in i18n-js's own official readme](https://github.com/fnando/i18n-js#pluralization).

Compared to that, i18next has [pluralization rules for cardinals, ordinals and something more in most languages](https://www.i18next.com/translation-function/plurals) ready to use, that means you can rightly pass on most the burden of dealing with language-specific logic to translators. I'm not devoted to i18next, and I think other competitors like [FormatJS](/formatjs/formatjs) or [Globalize](/globalizejs/globalize) can do as much the same thing in different ways, just i18n-js probably not.

Also one important feature i18n-js lacks is JSX integration that all JS libraries above seem to have. For example, you can do such kind of thing across the element boundary (see [this diff](https://github.com/forem/forem/pull/14932/files#diff-26ecb19806d86252f63bb99aa8633b996eb624a3b8f8abac96d69bbff1de3018)).

https://github.com/forem/forem/blob/2848d91def282afe2074769b0af91ef31a757d1a/app/javascript/articles/components/CommentsCount.jsx#L30-L34

with the English translation data looks like:

https://github.com/forem/forem/blob/2848d91def282afe2074769b0af91ef31a757d1a/app/javascript/i18n/en.json#L147-L148

while in Japanese:

https://github.com/forem/forem/blob/2848d91def282afe2074769b0af91ef31a757d1a/app/javascript/i18n/ja.json#L145

This will save you a lot of `dangerouslySetInnerHTML` I guess.

Reply:
FWIW, if I'm allowed to choose an idealistic translation library in JS, I'd go for [Fluent](/projectfluent/fluent.js). It is IMO the closest one to the translators' holy grail at this point, except it introduces a completely new format and concepts that may be difficult to users, and not very battle-proven yet. A library like i18next is more modest and adaptive to the "industry standard".

Reply:
Thanks for the comments, this is very insightful. FWIW, I'm totally on board with using I18next, I just wanted to see the benefits of one/the shortcoming of the other more clearly spelled out. I also appreciate your pragmatism, though I did bookmark Fluent for latter, it seems very interesting, thanks for the pointer!